src/noit_metric_director.c

```diff
@@ -137,15 +139,52 @@ noit_metric_director_my_lane() {
   return get_my_lane();
 }
 
+mtev_hook_return_t
+hook_metric_subscribe_account(void *closure, noit_metric_message_t *m, int *w, int wlen) {
+  int account_id = m->id.account_id;
+  int *account_interests;
+  if(mtev_hash_retrieve(&account_interests_map, (const char *)&account_id, sizeof(account_id), (void **)&account_interests)){
+    for(int i=0; i<wlen; i++){
+      if(w[i] == 0 && account_interests[i] > 0)
+        w[i] = 1;
+    }
+  }
+  return MTEV_HOOK_CONTINUE;
+}
+
+caql_cnt_t
+noit_adjust_account_interest(int account_id, short cnt) {
+  int thread_id, icnt, *account_interests = NULL, rc, *account_id_copy;
+  while(!mtev_hash_retrieve(&account_interests_map, (const char *)&account_id, sizeof(account_id), (void **)&account_interests)){
+    account_interests = calloc(nthreads, (sizeof(*account_interests)));
+    account_id_copy = calloc(1, sizeof(int));
+    *account_id_copy = account_id;
+    rc = mtev_hash_store(&account_interests_map, (const char *)account_id_copy, sizeof(*account_id_copy), (void **)account_interests);
+    if(!rc) { // Lost race. Try again.
+      free(account_id_copy);
+      free(account_interests);
+    }
+    else {
+      break;
+    }
+  }
+  mtevAssert(account_interests);
+  thread_id = get_my_lane();
+  icnt = account_interests[thread_id];
+  icnt += cnt;
+  if (icnt < 0) icnt = 0;
+  mtevAssert(icnt <= 0xffff);
+  account_interests[thread_id] = icnt;
```

Comment:
I read this file last Friday prior to this PR and I've seen this pattern in a few places already... and I don't understand how this is thread safe.  Nor can I can see how it possibly can be.

Can someone enlighten on this please?  What is keeping ``account_interests`` safe?  Because I do not see it.

Reply:
I believe it is not thread safe and it assumes that you can only ever operate on your thread.  If this assumption is not true, it is broken.

Reply:
@postwait 

It's a static variable in the presence of multiple threads.  Then it's not safe to use.  The assumption seems to be that it's safe because of the index accessed is based on ``thread_id``.

This is incorrect because of false sharing.  For this to work properly, we need at a minimum one cache line of space between the "thread slots"